HSBM.py

```diff
@@ -0,0 +1,442 @@
+from graspy.models.sbm import _calculate_block_p, _block_to_full, _get_block_indices
+from graspy.models.base import BaseGraphEstimator
+from graspy.utils import augment_diagonal, pass_to_ranks, import_graph, remove_loops, symmetrize
+
+from sklearn.utils.validation import check_is_fitted
+from sklearn.utils import check_array, check_X_y
+from sklearn.base import BaseEstimator
+from scipy.stats import chi2
+# from spherecluster import SphericalKMeans
+
+
+def _check_common_inputs(
+    min_components, max_components, cluster_kws, embed_kws
+):
+    if not isinstance(min_components, int):
+        raise TypeError("min_components must be an int")
+    elif min_components < 1:
+        raise ValueError("min_components must be > 0")
+
+    if not isinstance(max_components, int):
+        raise TypeError("max_components must be an int")
+    elif max_components < 1:
+        raise ValueError("max_components must be > 0")
+    elif max_components < min_components:
+        raise ValueError("max_components must be >= min_components")
+
+    if embed_kws:
+        if not isinstance(embed_kws, dict):
+            raise TypeError("embed_kws must be a dict")
+
+    if not isinstance(cluster_kws, dict):
+        raise TypeError("cluster_kws must be a dict")
+
+
+class RecursiveCluster(NodeMixin, BaseEstimator):
+    def __init__(
+        self,
+        selection_criteria=None,
+        cluster_method="GMM",
+        n_init=1,
+        parent=None,
+        min_components=1,
+        max_components=10,
+        cluster_kws={},
+        min_split=1,
+        max_level=20,
+        delta_criter=None,
+        likelihood_ratio=None,
+    ):
+
+        _check_common_inputs(
+            min_components, max_components, cluster_kws, embed_kws=None
+        )
+
+        if cluster_method not in ["GMM", "KMeans", "Spherical-KMeans"]:
+            msg = "clustering method must be one of {GMM, Kmeans, Spherical-KMeans}"
+            raise ValueError(msg)
+
+        if delta_criter:
+            if delta_criter < 0:
+                raise ValueError("delta_criter must be positive")
+
+        if likelihood_ratio:
+            if likelihood_ratio <= 0 or likelihood_ratio >= 1:
+                raise ValueError("likelihood_ratio must be in (0,1)")
+
+        self.parent = parent
+        self.n_init = n_init
+        self.min_components = min_components
+        self.max_components = max_components
+        self.cluster_method = cluster_method
+        self.min_split = min_split
+        self.selection_criteria = selection_criteria
+        self.cluster_kws = cluster_kws
+        self.min_split = min_split
+        self.max_level = max_level
+        self.delta_criter = delta_criter
+        self.likelihood_ratio = likelihood_ratio
+        
+    def fit(self, X, adj=None):
+        self.fit_predict(X)
+        return self
+
+    def predict(self, X, adj=None):
+        check_is_fitted(self, ["model_"], all_or_any=all)
+        labels = self.fit_predict(X)
```

Comment:
predict should not refit the model to the new data that was passed, should only run the prediction step. so will probably want to separate the predict functionality into this function or somewhere else, and then call it from `fit_predict`

Reply:
Oh I see what you're saying. Will fix this.